### ISSUE TITLE
How to Convert 2DFAN from Pytorch to Keras and to TensorFlow .pb?

### ISSUE THREAD

**Issue by felipheggaliza** [iperov/DeepFaceLab#195]
Hi,

Congratulations for building such a impressive tool!

I am studying this Face Aligment Network [FAN ](https://github.com/1adrianb/face-alignment) and also working in a side project were I would like to use this network inside TensorFlow. For that I need to generate a .pb file.

I manage to convert your [2DFAN-4.h5](https://github.com/iperov/DeepFaceLab/blob/master/facelib/2DFAN-4.h5)  to 2DFAN-4.pb. But I am struggling to make an inference. 

Could you explain me what tool did you use to go from Pytorch to Keras h5 model? That way I can take a detailed look at network, in order to understand what is the input and output tensor names.

This is my code for loading the .pb file:

```
def load_graph(frozen_graph_filename):
        # We load the protobuf file from the disk and parse it to retrieve the 
        # unserialized graph_def
        with tf.gfile.GFile(frozen_graph_filename, "rb") as f:
            graph_def = tf.GraphDef()
            graph_def.ParseFromString(f.read())

        # Then, we import the graph_def into a new Graph and returns it 
        with tf.Graph().as_default() as graph:
            # The name var will prefix every op/nodes in your graph
            # Since we load everything in a new graph, this is not needed
            tf.import_graph_def(graph_def, name="prefix")
        return graph
```

This is my code for running the prediction:

```
def predict(model, input_image):
        with tf.Session() as sess:
            
            input_layer = model.get_tensor_by_name('prefix/input_1:0') # Input Tensor
            output_layer = model.get_tensor_by_name('prefix/conv2d_194/convolution:0') 

            landmarks_heatmap = sess.run(output_layer, feed_dict = {input_layer : input_image} )
            
            return landmarks_heatmap
```

```
frozen_graph_filename = "2DFAN-4.pb"
model = load_graph(self, frozen_graph_filename)
2d_fan_output = predict(model , input_image)
```

**Comment by iperov** [comment]
I "manually" transferred weights from pytorch model to keras model by this code:
[port.zip](https://github.com/iperov/DeepFaceLab/files/2983341/port.zip)
also I changing from channels_first to channels_last requires no additional code.

I did same thing for MTCNN and S3FD models transfer.

**Comment by iperov** [comment]
why are you sticked to raw tensorflow impl? tf.keras is high level superset same as keras

**Comment by felipheggaliza** [comment]
Hi @iperov! 

Thank you very much for sharing the code, I belived it will help me to understand a little bit more about the FAN model!

One of my end goals is to run inference using the FAN model in a simple device, such as a simple Intel Core i7, one or multiple Raspberry Pi 3 machines. But since this is a heavy model, I will probably need to optimize it to run faster. I am thinking about ways to do both multiprocessing and also graph optimization such as INT8 Quantization using TensorFlow Lite or OpenVino.

TF-Lite: 
https://www.tensorflow.org/lite/guide/faq

OpenVino: 
https://docs.openvinotoolkit.org/R5/_docs_IE_DG_Int8Inference.html
https://www.embedded-vision.com/industry-analysis/blog/introducing-int8-quantization-fast-cpu-inference-using-openvino

**Comment by iperov** [comment]
i dont think that fan can be optimized without retrain. 
FAN is not realtime even on gtx1060, and produces ~10 fps.

why you need 68 landmarks detector on mobile devices?

**Comment by iperov** [comment]
may be you need just convex hull of face mask produced by 68 points?
